internal/transport/grpchttp2/framer.go

```diff
@@ -0,0 +1,252 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package grpchttp2 defines HTTP/2 types and a framer API and implementation.
+package grpchttp2
+
+import "golang.org/x/net/http2/hpack"
+
+// FrameType represents the type of an HTTP/2 Frame.
+// See [Frame Type].
+//
+// [Frame Type]: https://httpwg.org/specs/rfc7540.html#FrameType
+type FrameType uint8
+
+const (
+	FrameTypeData         FrameType = 0x0
+	FrameTypeHeaders      FrameType = 0x1
+	FrameTypeRSTStream    FrameType = 0x3
+	FrameTypeSettings     FrameType = 0x4
+	FrameTypePing         FrameType = 0x6
+	FrameTypeGoAway       FrameType = 0x7
+	FrameTypeWindowUpdate FrameType = 0x8
+	FrameTypeContinuation FrameType = 0x9
+)
+
+// Flag represents one or more flags set on an HTTP/2 Frame.
+type Flag uint8
+
+const (
+	FlagDataEndStream          Flag = 0x1
+	FlagDataPadded             Flag = 0x8
+	FlagHeadersEndStream       Flag = 0x1
+	FlagHeadersEndHeaders      Flag = 0x4
+	FlagHeadersPadded          Flag = 0x8
+	FlagHeadersPriority        Flag = 0x20
+	FlagSettingsAck            Flag = 0x1
+	FlagPingAck                Flag = 0x1
+	FlagContinuationEndHeaders Flag = 0x4
+)
+
+// Setting represents the id and value pair of an HTTP/2 setting.
+// See [Setting Format].
+//
+// [Setting Format]: https://httpwg.org/specs/rfc7540.html#SettingFormat
+type Setting struct {
+	ID    SettingID
+	Value uint32
+}
+
+// SettingID represents the id of an HTTP/2 setting.
+// See [Setting Values].
+//
+// [Setting Values]: https://httpwg.org/specs/rfc7540.html#SettingValues
+type SettingID uint16
+
+const (
+	SettingsHeaderTableSize      SettingID = 0x1
+	SettingsEnablePush           SettingID = 0x2
+	SettingsMaxConcurrentStreams SettingID = 0x3
+	SettingsInitialWindowSize    SettingID = 0x4
+	SettingsMaxFrameSize         SettingID = 0x5
+	SettingsMaxHeaderListSize    SettingID = 0x6
+)
+
+// FrameHeader is the 9 byte header of any HTTP/2 Frame.
+// See [Frame Header].
+//
+// [Frame Header]: https://httpwg.org/specs/rfc7540.html#FrameHeader
+type FrameHeader struct {
+	// Size is the size of the frame's payload without the 9 header bytes.
+	// As per the HTTP/2 spec, size can be up to 3 bytes, but only frames
+	// up to 16KB can be processed without agreement.
+	Size uint32
+	// Type is a byte that represents the Frame Type.
+	Type FrameType
+	// Flags is a byte representing the flags set on this Frame.
+	Flags Flag
+	// StreamID is the ID for the stream which this frame is for. If the
+	// frame is connection specific instead of stream specific, the
+	// streamID is 0.
+	StreamID uint32
+}
+
+// Frame represents an HTTP/2 Frame.
+type Frame interface {
+	Header() *FrameHeader
+	// Free frees the underlying buffer if present so it can be reused by the
+	// framer.
```

Comment:
I'm having second thoughts about using "Free" for this method. In line with Go's conventions, we typically "put" the underlying buffer back into the pool. 

Or the method should be called PutBuffer? 

@easwars: any thoughts?

Reply:
I would say the name doesn't matter much since we are going to remove this method from the interface when we integrate with the mem package.

I would recommend adding a todo here for the same.